src/solvers/refinement/string_refinement.cpp

```diff
@@ -1499,24 +1802,26 @@ static std::vector<exprt> sub_arrays(const exprt &array_expr)
 /// add to the index set all the indices that appear in the formula and the
 /// upper bound minus one
 /// \par parameters: a string constraint
-void string_refinementt::add_to_index_set(const exprt &s, exprt i)
+static void add_to_index_set(
+  std::map<exprt, std::set<exprt>> &index_set,
+  std::map<exprt, std::set<exprt>> &current_index_set,
+  const namespacet &ns,
+  const exprt &s, exprt i)
 {
   simplify(i, ns);
-  if(i.id()==ID_constant)
+  if(i.id()!=ID_constant || expr_cast<size_t>(i))
```

Comment:
to preserve the semantic `expr_cast<size_t>(i)` should be compared using `>= 0`

Reply:
`size_t` is unsigned, no need to do that comparison. If expression evaluates to something smaller than 0, it will return an empty optional, which will evaluate to false.

Reply:
I think that process is too implicit, I would appreciate if it was slightly longer, for example assigning `expr_cast<size_t>(i)` to a descriptive variable, e.g. `is_nonneg_size`.